avocado/core/nrunner.py

```diff
@@ -18,6 +23,224 @@
 #: runner that performs its work asynchronously
 RUNNER_RUN_STATUS_INTERVAL = 0.5
 
+#: All known runners
+KNOWN_RUNNERS = {}
+
+
+class SpawnMethod(enum.Enum):
+    """The method employed to spawn a runnable or task."""
+    #: Spawns by running executing Python code, that is, having access to
+    #: a runnable or task instance, it calls its run() method.
+    PYTHON_CLASS = object()
+    #: Spawns by running a command, that is having either a path to an
+    #: executable or a list of arguments, it calls a function that will
+    #: execute that command (such as with os.system())
+    STANDALONE_EXECUTABLE = object()
+    #: Spawns with any method available, that is, it doesn't declare or
+    #: require a specific spawn method
+    ANY = object()
+
+
+class BaseSpawner:
+    """Defines an interface to be followed by all implementations."""
+
+    METHODS = []
+
+    @staticmethod
+    def is_alive(task):
+        pass
+
+    def spawn(self, task):
+        pass
+
+
+class ProcessSpawner(BaseSpawner):
+
+    METHODS = [SpawnMethod.STANDALONE_EXECUTABLE]
+
+    @staticmethod
+    def is_alive(task):
+        return task.spawn_handle.returncode is None
+
+    @asyncio.coroutine
+    def spawn(self, task):
+        runner = pick_runner_command(task)
+        args = runner[1:] + ['task-run'] + task.get_command_args()
+        runner = runner[0]
+
+        #pylint: disable=E1133
+        task.spawn_handle = yield from asyncio.create_subprocess_exec(
+            runner,
+            *args,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE)
+
+
+class PodmanSpawner(BaseSpawner):
+
+    METHODS = [SpawnMethod.STANDALONE_EXECUTABLE]
+    IMAGE = 'fedora:31'
+
+    @staticmethod
+    def is_alive(task):
+        if task.spawn_handle is None:
+            return False
+
+        cmd = ["/usr/bin/podman", "ps", "--all", "--format={{.State}}",
```

Comment:
We are adding a new dependency here, right? Should from now on avocado depend on podman? If not, we should think on how to make this optional.

Reply:
I think we should at least have a graceful degradation mechanism. I don't think we need to enforce it being available, but if it's not available users should still have a pleasant experience. What do you think?

Reply:
Ok, no problem. But at least we should print a nice message saying that to use this, the user needs to install podman, right?

Reply:
Yes, exactly. That's my idea.

Reply:
@clebergnu sorry, I have edited the comment.

Reply:
> IIUC, the user will select the mechanism by a command-line option, like `--podman-spawner` or something like that.
> 
> So, if the user is selecting this, this should be used if installed and IMO we should not try to downgrade to any other mechanism other than podman (since that the user has explicitly selected podman)
> 
> So we have a few paths follow here:
> 
> 1. Make this optional, so when a plugin is installed this dependency will be also installed and the option will be available;
> 2. Make this part of the core, making the option `--podman-spawner` available even when the podman is not installed. So, we need to print a message if podman is not installed and the user is trying to use it.
> 3. Make this default (part of core) and force this dependency;
> 
> For me, it makes sense 1 (if podmand is not the default option) or 3 (if we are going to use podman as default to provide the isolation).
> 
> But I understand that we are trying to provide basic support here and improve this in the near future.
> 
> Do you have any other alternative? Maybe it was not clear to me your plan.

I think it's completely reasonable to have an option that will make use of an external feature, provided that feature is installed on the system. So I'm going with option 2 here without any regrets. Doing more than that is way more complex and I don't see any extra value.

Reply:
ok, no problem.

Reply:
What I've done is this:

```
avocado nrun --podman-spawner /bin/true
Podman Spawner selected, but podman binary "/usr/bin/podman" is not available on the system.  Please install podman before attempting to use this feature.
```